app/src/main/java/com/github/se/eventradar/model/repository/event/FirebaseEventRepository.kt

```diff
@@ -80,4 +83,40 @@ class FirebaseEventRepository : IEventRepository {
       Resource.Failure(e)
     }
   }
+
+  override fun observeEvents(): Flow<Resource<List<Event>>> = callbackFlow { // to be tested
+    val listener =
+        eventRef.addSnapshotListener { snapshot, error ->
+          if (error != null) {
+            trySend(
+                Resource.Failure(Exception("Error listening to event updates: ${error.message}")))
+            return@addSnapshotListener
+          }
+          val events = snapshot?.documents?.mapNotNull { Event(it.data!!, it.id) }
+          trySend(Resource.Success(events ?: listOf()))
+        }
+    awaitClose { listener.remove() }
+  }
+
+  override fun observeUpcomingEvents(userId: String): Flow<Resource<List<Event>>> = callbackFlow {
+    val query = eventRef.whereArrayContains("attendees_list", userId)
```

Comment:
Both your `observers` are good! Small detail for the future: you could cut the number of db calls in half. Observe upcoming events is in fact the list of all events filtered (i.e. dont need to receive the list twice)

You can see example number 1 and 2 in [this file](https://github.com/RaulinN/calculator-compose/blob/master/app/src/main/java/com/github/raulinn/calculator/CalculatorViewModel.kt) that I shared previously to answer the question about cognitive complexity